### ISSUE TITLE
Take the iterated accordion component to the accessibility clinic

### ISSUE THREAD

**Issue by hannalaakso** [alphagov/govuk-frontend#2295]
## What
Take iterated component to accessibility clinic.

## Why
To get expert feedback

## Who needs to know about this
- GOV.UK Accessibility team
- Design System developers, designers

## Done when
- [x] Book session
- [x] Prepped a discussion plan for the session
- [x] Arranged a notetaker for the session
- [x] Attended the session
- [x] Recorded feedback

## Further detail
List of things to discuss here: https://docs.google.com/document/d/1CG0OUT1tzy7q9juK4O0Iv974854hiBK5GMGF9_0ma08/edit#heading=h.8n015y3xrh1t

Previous accessibility clinic findings:
https://github.com/alphagov/govuk-design-system/issues/1706#issuecomment-865908694

**Comment by hannalaakso** [comment]
No clinics in August. Have sent an email to accessibility clinic to ask about being put on a waiting list.

**Comment by hannalaakso** [comment]
We've got a 30 minute session for 26th August 🎉 

We'd like to cover
- [the Firefox High Contrast mode behaviour](https://github.com/alphagov/govuk-frontend/issues/2321)
- testing the accordion with Dragon Naturally Speaking
- whether there are any potential issues with having the heading, summary and icon button inside a `<button>`
- show how the accordion presents when CSS is disabled but JS isn't (this is a real edge case but it would still be good to get an opinion if there's time)

**Comment by hannalaakso** [comment]
We had some existing issues to address some of the above actions, I’ve tweaked them slightly to include this latest clinic feedback:
- https://github.com/alphagov/govuk-frontend/issues/2321
- https://github.com/alphagov/govuk-frontend/issues/2327

Created a new card to potentially iterate the design of the summary line
- https://github.com/alphagov/govuk-frontend/issues/2336

I'm closing this issue as completed as we'll review the clinic feedback again in our catchup to see if any more cards need to be raised.

**Comment by hannalaakso** [comment]
# Solving accessibility issues flagged by the clinic

Following the [findings from the accessibility clinic,](https://github.com/alphagov/govuk-frontend/issues/2295#issuecomment-906449543) the Design System team have since sought feedback from the clinic on **two alternative prototypes where the summary line is placed differently in the markup**. 

The visual design didn’t change in these prototypes.

## Prototype 1: The heading/button only contains the heading text

In this alternative prototype, the `<h2>` and `<button>` only contained the heading text. The summary and the “Show” text were in separate divs below. 

The summary and the “Show” text were connected to the heading/button with `aria-describedby`.  

![Screenshot 2021-09-20 at 11 19 38](https://user-images.githubusercontent.com/5007934/133987339-a24de724-5bc9-4d7b-b4a9-3bb856f41685.png)

### Feedback from accessibility team on prototype 1

The feedback from the accessibility clinic was that this most likely would be a WCAG fail since the heading looks like a heading and not a button so this could confuse users since screenreaders would announce the elements differently than is indicated by the visual design, as demonstrated by the following screen grab:

![Screenshot 2021-09-20 at 11 23 55](https://user-images.githubusercontent.com/5007934/133987689-20bc5710-c34d-4915-be37-92674f4ad282.png)

## Prototype 2: The button only contains the “Show” text
In this prototype, `<button>` only contained the “Show” text. `<h2>` only contained the heading text and the summary was in its own div below. The heading was connected to the button with `aria-describedby` on the button 

![Screenshot 2021-09-20 at 11 29 09](https://user-images.githubusercontent.com/5007934/133988319-61b458f3-1156-441b-9fc2-7d72c5661a28.png)

The summary line was deemed secondary information by the accessibility team so we followed their advice by not connecting it to the button.

From our team’s point of view, this prototype would need user testing to find out whether sighted users understand the focus state when only “Show” has a visual focus state, as shown below:

![Screenshot 2021-09-20 at 11 29 40](https://user-images.githubusercontent.com/5007934/133988364-74bb15d3-85f2-4e9b-8367-1618ac9ef861.png)

This diverges from the new GOV.UK Design which has a visual focus state on the heading, summary and “Show”, and the current Design System version shows a focus state on the heading text.

Another potential issue, noted when our team tested prototype 2 with screen readers, was that the reader mode in both Jaws and NVDA treats the heading text and button as separate elements when the user navigates the page which is different from the behaviour in the current Design System version and the new GOV.UK Design.

### Feedback from accessibility team on prototype 2
The feedback from the accessibility team was whilst this version was semantically robust (the button only contains the call to action text), it would be changing the touch area to only cover the “Show” text; this could be a usability issue since we know from research that users often click on the heading to open the accordion (this is how accordions typically work). If we made the touch area to cover the whole section header to fix this (but still only have the “Show” inside the button), it could be a WCAG fail (further confirmation on this would be needed) since the heading wouldn’t be available to keyboard users as an interactive element (unlike mouse users who could click anywhere in the section header). Also with users who change colours in their browser might see the button element only around “Show” even though the whole section header would actually interactive. More user research would be needed to understand how well this solution would work for users.

More user research would also be needed to understand whether separating the heading and button text in Jaws and NVDA reader mode (as discussed above) could be an issue for users.

Prototype 2 could be improved with an `aria-label` that contains `[heading text][“Show this section”]` on the button, since using `aria-describedby` to connect the heading makes the heading text read out quite late (after "Show this section" and information about the button element) in screen reader announcements. We could add thus `aria-label` content in JavaScript to avoid asking teams to duplicate the information in their markup. 

# What we have learnt and next steps

## What causes the issue

**The core issue here seems to be that:**
1. **[The new GOV.UK design for the accordion](https://github.com/alphagov/govuk-frontend/pull/2257) includes more textual information about how the accordion works (“Show this section”), both to help to solve the [known accessibility issue](https://design-system.service.gov.uk/accessibility/#non-compliance-with-the-accessibility-regulations) but also to help to improve the general accessibility of the component since we know that [not all users notice accordions or understand how they work](https://design-system.service.gov.uk/components/accordion/#when-not-to-use-this-component); the second point was one of the main messages from the Working Group when they [recently reviewed the accordion](https://github.com/alphagov/govuk-design-system-backlog/issues/1#issuecomment-839802842).** 
2. **The accordion is a complex component with many moving parts, even before the inclusion of the summary line. This design change to include more text in the core design has made the summary line text problematic by being yet another piece of information that needs to be announced to users when they interact with the accordion.**

## We have confidence that the new GOV.UK accordion works for users
The [new version of the accordion designed by GOV.UK](https://github.com/alphagov/govuk-frontend/pull/2257) has been in use on GOV.UK since earlier this year (although only the Service Manual is using the version with the summary line) so we have sufficient confidence that it works for users. The new version, without the summary line, has also undergone lab user research and no issues were observed.

The new design was agreed by the accessibility team to be an improvement on the existing Design System version as it solves:
- the existing WCAG fail which we need to fix
- the other existing issue listed in the current guidance about users potentially missing the + icon when zooming in

## The inclusion of the summary line could be an issue for users
The accessibility clinic advises that they can’t say whether including the summary line inside the button is definitely an issue for users, but it also doesn’t seem ideal to include so much content in the button. More user research would be needed to establish whether it could confuse users. 

The problem definitely appears worse when both the heading and the summary are very long, as in the below screenshot:

![Screenshot 2021-09-20 at 11 55 19](https://user-images.githubusercontent.com/5007934/133991257-62fc1d37-3804-4313-b1fa-9daf22f2cf7e.png)

##  How we're going to mitigate potential issues with the summary line

To help mitigate potential issues with the summary line:
- we're going to change the [Design System guidance](https://design-system.service.gov.uk/components/accordion/) to advise against using the summary line in most situations
- where the summary line is used, we will advise that the heading and summary should contain **six words in total** 
- we will encourage teams to look out for any issues in research and user feedback to help us iterate the guidance if necessary in the future

**Comment by hannalaakso** [comment]
I've moved this back into the Backlog to remind us to review whether the accessibility clinic should take one last look at the component before publishing it.

**Comment by hannalaakso** [comment]
Last couple of bits:
- Dragon testing

**Comment by vanitabarrett** [comment]
## Notes from the accessibility clinic

### Dragon testing
- No problems observed.
- Tested ’show/hide all sections’; click ‘Vaccination’ (accordion heading); ‘click show, choose 1’; click ‘show all’; ‘click vaccination show/hide’. Tested in Chrome and IE11. Also tested with the summary line version in IE11 and it works as expected.

### Update on what we'll put in our guidance re: summary lines
- Not deprecating, recommending against using it. Avoid using it if you can, as it makes the content unnecessarily complicated. 
- Do we want to say it will be taken out in the future? We want to keep things as flexible as possible. We don’t know for sure that it’s a problem. We don’t have lots of evidence.

### Mention the semantics of buttons now
- Added a second comma into the button. Comma after the heading, and also after the summary text. Helps divide things up a little bit. The comma is visually hidden.
- What about if there is a comma in the summary text? Probably not making things worse though.
- Comma not just for screenreaders, also for Google search results - might want to make a note that this change helps with that. Anika said that the new version ‘show this section’ will appear in search results now? Hanna mentioned there might be a blog post by GOV.UK (maybe Andy Sellick) about using a visually hidden comma to break things up - maybe about step by step?
- Anika - overall the changes improve more than they make things worse
- Anika noticed the text with CSS disabled but JS enabled - some of the text appears larger. Not a concern, but a bit weird. Assumed it was because it has the govuk-body class, but that can’t be the cause if we’ve disabled CSS. This was in Chrome. Anika says it looks ok to her when checking in her browser (also Chrome). This could have been an earlier screenshot? Hanna also can't seem to replicate now.

### Heads up on NVDA browse mode 
- More just a heads up but we found this bug Accordion NVDA browse mode adjustment causes long announcements to be cut off in NVDA browse mode #2373 and ended up removing the NVDA adjustment
- Text being cut off, have to press down key to read the next ‘bit’. But we were seeing the line getting clipped, but when you pressed down it never read out the rest of the line.
- Pseudo element had been added to make inline elements appear block. Taken it out to stop the clipping of the announcements.

## Potential Actions

- Find the blog post that GOV.UK wrote about adding a visually hidden separator. We think this might have been a blog post by Andy Sellick about the step by step nav.
- We could investigate the text styling issue when CSS is disabled but JS is running - probably not worth spending much time on it, if any
- Anika mentioned our changes might mean the show/hide button now appears in Google search results - do we need to confirm this and do something about it and/or document that change?

**Comment by hannalaakso** [comment]
Closing this as the work has been completed.

I've made a note for myself to create a card to look at the non-indexing of the 'Show' text as discussed above, will link to it from here when done.
